### PR TITLE
[MNG-7648] Fix locationTracking in DefaultModelBuildingRequest copy constructor.

### PR DESCRIPTION
**Pull request by pavelhoral**
This fixes missing `locationTracking` in project model.

Missing `locationTracking` was causing issues down the line in enforcer plugin that was not able to verify source of the model configuration.

--

Following this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue. Your pull request should address just this issue, without
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] SUMMARY`,
       where you replace `MNG-XXX` and `SUMMARY` with the appropriate JIRA issue.
 - [x] Also format the first line of the commit message like `[MNG-XXX] SUMMARY`.
       Best practice is to use the JIRA issue title in both the pull request title and in the first line of the commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Comment by pavelhoral** [comment]
I can verify that with the fixed version I am able to successfully build my test project (i.e. this fix fixes problems in enforcer):

```console
$ ../apache-maven-4.0.0-alpha-4-SNAPSHOT/bin/mvn -s ../settings.xml verify -f project/pom.xml
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------------------------< org.example:enforcer-test-project >-------------------------------------------
[INFO] Building enforcer-test-project 1.0.0-SNAPSHOT
[INFO]   from pom.xml
[INFO] ---------------------------------------------------------[ pom ]----------------------------------------------------------
[INFO] 
[INFO] --- enforcer:3.1.1-SNAPSHOT:enforce (enforce-plugin-versions) @ enforcer-test-project ---
[INFO] Executing rule: org.apache.maven.plugins.enforcer.RequirePluginVersions
[INFO] --------------------------------------------------------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] --------------------------------------------------------------------------------------------------------------------------
[INFO] Total time:  0.552 s
[INFO] Finished at: 2022-12-27T12:21:02+01:00
[INFO] --------------------------------------------------------------------------------------------------------------------------
```

Previously:

```console
$ ../apache-maven-4.0.0-alpha-3/bin/mvn -s ../settings.xml verify -f project/pom.xml
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------------------------------------< org.example:enforcer-test-project >-------------------------------------------
[INFO] Building enforcer-test-project 1.0.0-SNAPSHOT
[INFO]   from pom.xml
[INFO] ---------------------------------------------------------[ pom ]----------------------------------------------------------
[INFO] 
[INFO] --- enforcer:3.1.1-SNAPSHOT:enforce (enforce-plugin-versions) @ enforcer-test-project ---
[INFO] Executing rule: org.apache.maven.plugins.enforcer.RequirePluginVersions
[ERROR] Rule 0: org.apache.maven.plugins.enforcer.RequirePluginVersions failed with message:
Some plugins are missing valid versions or depend on Maven 4.0.0-alpha-3 defaults: (LATEST RELEASE SNAPSHOT are not allowed)
org.apache.maven.plugins:maven-clean-plugin.    The version currently in use is 3.2.0 via super POM or default lifecycle bindings
org.apache.maven.plugins:maven-wrapper-plugin.  The version currently in use is 3.1.1 via super POM or default lifecycle bindings
org.apache.maven.plugins:maven-install-plugin.  The version currently in use is 3.0.1 via super POM or default lifecycle bindings
org.apache.maven.plugins:maven-site-plugin.     The version currently in use is 4.0.0-M4 via super POM or default lifecycle bindings
org.apache.maven.plugins:maven-deploy-plugin.   The version currently in use is 3.0.0 via super POM or default lifecycle bindings
org.apache.maven.plugins:maven-enforcer-plugin.         The version currently in use is 3.1.1-SNAPSHOT via super POM or default lifecycle bindings

[INFO] --------------------------------------------------------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] --------------------------------------------------------------------------------------------------------------------------
[INFO] Total time:  0.573 s
[INFO] Finished at: 2022-12-27T12:22:15+01:00
[INFO] --------------------------------------------------------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.1.1-SNAPSHOT:enforce (enforce-plugin-versions) on project enforcer-test-project: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the '-e' switch
[ERROR] Re-run Maven using the '-X' switch to enable verbose output
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Review by slawekjaranowski — COMMENTED** [review]
Looks ok for me.

We should add a test for it, maybe in `org.apache.maven.project.ProjectBuilderTest`

**Review by mthmulders — APPROVED** [review]

**Comment by pavelhoral** [comment]
@slawekjaranowski I have created unit test for location tracking but I have kept it in a separate branch for now - https://github.com/pavelhoral/maven/commit/c25b984b24e46985a45f034535b8030437355642 . Let me know if this is something you had in mind and if I should add it to this PR or keep it separate.

**Comment by slawekjaranowski** [comment]
Please add test to this PR - we will have a complete change.

**Comment by pavelhoral** [comment]
I have added the unit test to this PR and I have also fixed `assertNotNull` argument order in the test next to the one I have added.

**Review by gnodet — APPROVED** [review]

File: maven-core/src/test/java/org/apache/maven/project/ProjectBuilderTest.java [hunk 1]
```diff
@@ -347,4 +349,32 @@ public void testBuildFromModelSourceResolvesBasedir() throws Exception {
         }
         assertEquals(0, errors);
     }
+
+    @Test
+    public void testLocationTrackingResolution() throws Exception {
+        File pom = getProject("MNG-7654");
```
Comment (slawekjaranowski) [inline maven-core/src/test/java/org/apache/maven/project/ProjectBuilderTest.java:355]: Should be `MNG-7648` - 7654 it is not issue for this PR.

Comment (pavelhoral) [reply maven-core/src/test/java/org/apache/maven/project/ProjectBuilderTest.java:355]: OMG how have I switched that :facepalm:

Comment (pavelhoral) [reply maven-core/src/test/java/org/apache/maven/project/ProjectBuilderTest.java]: Fixed (such a stupid mistake).

**Review by slawekjaranowski — APPROVED** [review]

**Comment by slawekjaranowski** [comment]
@pavelhoral thanks

I squashed to one final commit 

In 3.9.x there is no such issue - looks like `DefaultModelBuildingRequest` constructor is not used
but I think that we can cherry pick this commit to be sure that is working as expected with next changes

**Comment by pavelhoral** [comment]
> In 3.9.x there is no such issue - looks like `DefaultModelBuildingRequest` constructor is not used but I think that we can cherry pick this commit to be sure that is working as expected with next changes

@slawekjaranowski Should I create another PR with the unit test targeting `maven-3.9.x` branch? Or was that comment meant in a way that you will do that?

**Comment by slawekjaranowski** [comment]
@pavelhoral please create PR for `maven-3.9.x`

**Comment by hazendaz** [comment]
@pavelhoral Do you know if alpha 5 has this regression again?  Testing alpha 4 here https://github.com/hazendaz/spotbugs-maven-plugin/actions/workflows/it-maven-4.0.0.yaml and it works, the prior runs for same job used alpha 5 and seems broken again.  Just wonder if you are seeing same behaviour.

**Comment by gnodet** [comment]
@pavelhoral I'm planning a release of 4.0.0-alpha-6 in the coming days, so a quick check with the latest master branch would be welcomed !

**Comment by pavelhoral** [comment]
> @pavelhoral Do you know if alpha 5 has this regression again? Testing alpha 4 here https://github.com/hazendaz/spotbugs-maven-plugin/actions/workflows/it-maven-4.0.0.yaml and it works, the prior runs for same job used alpha 5 and seems broken again. Just wonder if you are seeing same behaviour.

It shouldn't be the same bug as that would have been caught by the unit tests (I hope).

Interesting thing is that the enforcer rule fails only on Java 17 and not Java 11:

<details><summary>Java 11 run</summary>

```console
/tmp/spotbugs-maven-plugin$ JAVA_HOME=/usr/lib/jvm/java-11-openjdk-amd64/ /srv/repo/maven/apache-maven-4.0.0-alpha-5/bin/mvn  -B -V clean install -Dlicense.skip=true -Dmaven.min-version=4.0.0-alpha-5 --no-transfer-progress
Apache Maven 4.0.0-alpha-5 (26d10a4bf9a2df75feef60da01d8706f2bf77a47)
Maven home: /srv/repo/maven/apache-maven-4.0.0-alpha-5
Java version: 11.0.19, vendor: Ubuntu, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.19.0-43-generic", arch: "amd64", family: "unix"
[INFO] Scanning for projects...
[INFO] 
[INFO] --------------------------------------< com.github.spotbugs:spotbugs-maven-plugin >---------------------------------------
[INFO] Building spotbugs-maven-plugin 4.7.3.6-SNAPSHOT
[INFO]   from pom.xml
[INFO] -----------------------------------------------------[ maven-plugin ]-----------------------------------------------------
[INFO] 
[INFO] --- enforcer:3.3.0:enforce (enforce-clean) @ spotbugs-maven-plugin ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.dependency.BannedDependencies passed
[INFO] Rule 1: org.apache.maven.enforcer.rules.BanDuplicatePomDependencyVersions passed
[INFO] Adding ignore: module-info
[INFO] Rule 2: org.codehaus.mojo.extraenforcer.dependencies.EnforceBytecodeVersion passed
[INFO] Rule 3: org.apache.maven.enforcer.rules.ReactorModuleConvergence passed
[INFO] Rule 4: org.apache.maven.enforcer.rules.version.RequireJavaVersion passed
[INFO] Rule 5: org.apache.maven.enforcer.rules.version.RequireMavenVersion passed
[INFO] Rule 6: org.apache.maven.enforcer.rules.RequirePluginVersions passed
[INFO] Rule 7: org.apache.maven.plugins.enforcer.RestrictImports passed
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ spotbugs-maven-plugin ---
[INFO] Deleting /tmp/spotbugs-maven-plugin/target (includes = [**/*], excludes = [])
[INFO]
...
```
</details>

<details><summary>Java 17 run</summary>

```console
/tmp/spotbugs-maven-plugin$ $ JAVA_HOME=/usr/lib/jvm/java-17-openjdk-amd64/ /srv/repo/maven/apache-maven-4.0.0-alpha-5/bin/mvn  -B -V clean install -Dlicense.skip=true -Dmaven.min-version=4.0.0-alpha-5 --no-transfer-progress
Apache Maven 4.0.0-alpha-5 (26d10a4bf9a2df75feef60da01d8706f2bf77a47)
Maven home: /srv/repo/maven/apache-maven-4.0.0-alpha-5
Java version: 17.0.7, vendor: Private Build, runtime: /usr/lib/jvm/java-17-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.19.0-43-generic", arch: "amd64", family: "unix"
[INFO] Scanning for projects...
[INFO] 
[INFO] --------------------------------------< com.github.spotbugs:spotbugs-maven-plugin >---------------------------------------
[INFO] Building spotbugs-maven-plugin 4.7.3.6-SNAPSHOT
[INFO]   from pom.xml
[INFO] -----------------------------------------------------[ maven-plugin ]-----------------------------------------------------
[INFO] 
[INFO] --- enforcer:3.3.0:enforce (enforce-clean) @ spotbugs-maven-plugin ---
[INFO] Rule 0: org.apache.maven.enforcer.rules.dependency.BannedDependencies passed
[INFO] Rule 1: org.apache.maven.enforcer.rules.BanDuplicatePomDependencyVersions passed
[INFO] Adding ignore: module-info
[INFO] Rule 2: org.codehaus.mojo.extraenforcer.dependencies.EnforceBytecodeVersion passed
[INFO] Rule 3: org.apache.maven.enforcer.rules.ReactorModuleConvergence passed
[INFO] Rule 4: org.apache.maven.enforcer.rules.version.RequireJavaVersion passed
[INFO] Rule 5: org.apache.maven.enforcer.rules.version.RequireMavenVersion passed
[INFO] Rule 7: org.apache.maven.plugins.enforcer.RestrictImports passed
[INFO] --------------------------------------------------------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] --------------------------------------------------------------------------------------------------------------------------
[INFO] Total time:  1.262 s
[INFO] Finished at: 2023-06-20T08:09:42+02:00
[INFO] --------------------------------------------------------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.3.0:enforce (enforce-clean) on project spotbugs-maven-plugin: 
[ERROR] Rule 6: org.apache.maven.enforcer.rules.RequirePluginVersions failed with message:
[ERROR] Some plugins are missing valid versions or depend on Maven 4.0.0-alpha-5 defaults (LATEST, RELEASE, SNAPSHOT, TIMESTAMP SNAPSHOT as plugin version are not allowed)
[ERROR]    org.apache.maven.plugins:maven-compiler-plugin.      The version currently in use is 3.11.0 via super POM or default lifecycle bindings
[ERROR] [ERROR] Best Practice is to always define plugin versions!
```
</details>


So I am wondering if the issue is that you are defining maven-compiler-plugin in jdk17on profile without the version: 

https://github.com/hazendaz/base-parent/blob/5a57310ad94bbae655ea412e1c9e8a3c66abfec5/pom.xml#L1647-L1650

I guess that plugin-management definitions are not merged (i.e. the version is not inherited from the other `pluginManagement/plugins` definition).

**Comment by hazendaz** [comment]
As far as I know it should be inherited.  At least it would not make sense to redefined versions all over.

**Comment by hazendaz** [comment]
Reporting again here, alpha 7 has same issue.  Only Alpha 4 of maven works as expected.  Any maven version before 4-alpha-5 and all 3.x as far as I can remember - like 10 years ago, do not require the version to be additional restated inside a profile.  Its always properly inherited it as far as I'm aware.  As @pavelhoral noted possible plugin definitions are not merged.  I presume they always were before, can anyone confirm?  Trying to understand why now after decade of doing profiles like this, its only an issue now.

**Comment by hazendaz** [comment]
You can see the issue additionally here https://github.com/spotbugs/spotbugs-maven-plugin/pull/626.  Its using same base parent and maven 3.3.9 and 3.9.4 are working fine.  If I use 4 alpha 4 its fine.  Only a problem on 4 afterwards.

**Comment by hazendaz** [comment]
```
 [INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.3.0:enforce (enforce-clean) on project spotbugs-maven-plugin: 
Error:  Rule 6: org.apache.maven.enforcer.rules.RequirePluginVersions failed with message:
Error:  Some plugins are missing valid versions or depend on Maven 4.0.0-alpha-5 defaults (LATEST, RELEASE, SNAPSHOT, TIMESTAMP SNAPSHOT as plugin version are not allowed)
Error:     org.apache.maven.plugins:maven-compiler-plugin. 	The version currently in use is 3.11.0 via super POM or default lifecycle bindings
Error: ] Best Practice is to always define plugin versions!
Error:  -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the '-e' switch
Error:  Re-run Maven using the '-X' switch to enable verbose output
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error: Process completed with exit code 1.
```

**Comment by hazendaz** [comment]
And yes my super pom - not maven's defines it in dependency management at 3.11.0.  So clearly something is wrong here.

**Comment by gnodet** [comment]
> And yes my super pom - not maven's defines it in dependency management at 3.11.0. So clearly something is wrong here.

Could your raise a new issue so that it's not forgotten ?

**Comment by hazendaz** [comment]
Yes will do.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jira-importer** [comment]
Resolve #8804

**Pull request merged**